[WEBSERVER] Handle HEAD requests in exactly the same way as GET requests

Karlson2k opened the pull request:
I started this PR as a fix for https://github.com/xbmc/xbmc/issues/20759

While fixing duplicated `Content-Length:` headers, I discovered that Kodi uses libmicrohttpd (MHD) API incorrectly. Responses for `HEAD` requests are formed as responses with zero-length body. This is incorrect.
MHD API uses the same responses for `GET` and `HEAD` requests. When response is used to answer `GET` request, it is completely sent, including body. When the same response is used to answer `HEAD` request, response body is not sent (even if body has non-zero length).

According to HTTP RFCs, reply to `HEAD` requests should be exactly the same as reply for the `GET` requests, but reply body must not be sent.

In this PR I fixed improper use of MHD API.

This PR is fixing several things:
* Response for `HEAD` requests are handled in exactly the same way as responses for `GET` requests (as per RFC), now response headers for `HEAD` request are **exactly** the same as in response for similar `GET` request.
* Fixed wrong `Content-Length: 0` header always present in replies for `HEAD` request. This may fix a number of problems in clients that use Kodi HTTP interface.
* Fixed duplicated `Content-Length:` headers in all responses.
* Unrelated small optimisation suggested by @lrusak 
* Fixed run-time warning in log at startup.
* Fixed order of initialization parameters to avoid run-time warning at startup.

lrusak commented:
When reviewing it's helpful to disable whitespace changes in some cases. In the case of this PR it reduces the amount of lines changes quite a lot. You can see it here https://github.com/xbmc/xbmc/pull/21019/files?w=1

|whitespace|additions|deletions|
|---|---|---|
|with|+92|−126|
|witout|+4|−38|

lrusak commented on xbmc/network/WebServer.cpp:
```diff
@@ -1425,12 +1403,9 @@ MHD_RESULT CWebServer::AddHeader(struct MHD_Response* response,
-    MHD_set_response_options(response, MHD_RF_INSANITY_HEADER_CONTENT_LENGTH, MHD_RO_END);
-  }
-#endif
+    m_logger->warn("Attempt to override MHD automatic \"Content-Length\" header");
```
Would be nice if this change was split into a separate commit as it seems to change the behaviour (logical commits, etc). I'm not really that worried about it though.

Karlson2k replied:
Makes sense. Separated changes.

lrusak commented on xbmc/network/WebServer.cpp:
```diff
@@ -827,112 +822,98 @@ MHD_RESULT CWebServer::CreateFileDownloadResponse(
-                                                    context->ranges);
-    }
+  uint64_t totalLength = 0;
+  std::unique_ptr<HttpFileDownloadContext> context(new HttpFileDownloadContext());
```
would be nice to update this to use make_unique but not required.
```suggestion
  std::unique_ptr<HttpFileDownloadContext> context = std::make_unique<HttpFileDownloadContext>();
```

Karlson2k replied:
Applied.

lrusak reviewed (commented):
Seems like the right thing to do. The only thing I'll add is that should our tests be amended to support the new behaviour of HEAD requests?

Karlson2k committed:
[WEBSERVER] Handle HEAD requests in exactly the same way as GET requests

According to HTTP RFCs, reply to HEAD requests should be the same as
reply for the GET requests, only reply body must not be sent.
This commit also fixed wrong "Content-Length: 0" header in replies for
HEAD requests.

Karlson2k committed:
[WEBSERVER] Fixed duplicated "Content-Length" headers

This commit fixed attempt to set redundant "Content-Length" header as
this header is always set by MHD when allowed by RFC.

Closes: https://github.com/xbmc/xbmc/issues/20759

Karlson2k committed:
[WEBSERVER] Minor optimisation: use `std::make_unique`

As suggested by @lrusak

Karlson2k commented:
@lrusak Applied all of your suggestions. Also updated PR description to make it clearer.

Karlson2k commented:
> Seems like the right thing to do. The only thing I'll add is that should our tests be amended to support the new behaviour of HEAD requests?

Current set of tests passed just fine with this changes. Do you think it's needed to add some additional tests?

Karlson2k commented:
I'm working on additional checks in the tests.

Karlson2k committed:
[WEBSERVER] Added checks in kodi-test

Karlson2k commented:
Added checks in tests.
Checked that not more than one `Content-Length` header is present or exactly one Content-Length` header is present (in some tests).
So it is not an "adaptation" of the tests, it is more check that bug from https://github.com/xbmc/xbmc/issues/20759 is fixed.

garbear reviewed (approved):
Welcome back!!! No problems that I can see, so +1

Karlson2k committed:
[WEBSERVER] Fixed run-time warning in log

`MHD_OPTION_EXTERNAL_LOGGER` should be the first parameter if used,
otherwise MHD may print initialisation errors by standard logger (stderr)
Latest MHD version warns if `MHD_OPTION_EXTERNAL_LOGGER` is used as not
the first parameter.

lrusak reviewed (approved):
Thank you for providing additional tests and going the extra mile to make some other improvements. 👍
